pallets/interest-accrual/src/lib.rs

```diff
@@ -204,35 +205,56 @@ pub mod pallet {
 	#[pallet::genesis_build]
 	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
 		fn build(&self) {
-			StorageVersion::<T>::put(Release::V1);
+			StorageVersion::<T>::put(Release::V2);
 		}
 	}
 
 	#[pallet::hooks]
 	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
 		fn on_initialize(_: T::BlockNumber) -> Weight {
-			let mut count = 0;
 			let then = LastUpdated::<T>::get();
 			let now = Self::now();
 			LastUpdated::<T>::set(Self::now());
 			let delta = Self::now() - then;
 			let bits = Moment::BITS - delta.leading_zeros();
-			Rate::<T>::translate(|per_sec, mut rate: RateDetailsOf<T>| {
-				count += 1;
-				Self::calculate_accumulated_rate(per_sec, rate.accumulated_rate, then, now)
+
+			// reads: timestamp, last updated, rates vec
+			// writes: last updated, rates vec
+			let mut weight = T::DbWeight::get().reads_writes(3, 2);
+
+			let rates = Rates::<T>::get();
+			let rates: Vec<_> = rates
+				.into_iter()
+				.filter_map(|rate| {
+					weight = weight.saturating_add(T::Weights::calculate_accumulated_rate(bits));
+
+					let RateDetailsOf::<T> {
+						interest_rate_per_sec,
+						accumulated_rate,
+						reference_count,
+					} = rate;
+
+					Self::calculate_accumulated_rate(
+						interest_rate_per_sec,
+						accumulated_rate,
+						then,
+						now,
+					)
 					.ok()
-					.map(|new_rate| {
-						rate.accumulated_rate = new_rate;
-						rate
+					.map(|accumulated_rate| RateDetailsOf::<T> {
+						interest_rate_per_sec,
+						accumulated_rate,
+						reference_count,
 					})
-			});
-
-			let db_weight = T::DbWeight::get().reads_writes(2, 1);
-
-			let db_weight_accumulated = T::DbWeight::get().reads_writes(1, 1)
-				+ T::Weights::calculate_accumulated_rate(bits);
+				})
+				.collect();
 
-			db_weight + db_weight_accumulated.saturating_mul(count)
+			Rates::<T>::set(
```

Comment:
I agree that in this case using `get` and `set` provides better readability than wrapping the `filter_map` inside `try_mutate`.

Reply:
I agree too, but here I recognize that by using `mutate()` we can avoid calling `collect()` and then avoid allocating a new big vector in memory. 🤔

Reply:
I don't think we can avoid the call to `collect` - part of what's happening with the `filter_map` is that any rates which overflow get removed (which will later cause extrinsics using those rates to fail). This isn't the _best_ error handling situation. Since there is no way to return an error from `on_initialize`, it's better to just nuke any bad data rather than allowing it to be used later on and further corrupt results

Reply:
Agreeing with Branan here!"

Reply:
You're right! I didn't pay attention to the filter 👍🏻, (although maybe `Vec::drain()` could do the trick?)